numpy/core/src/umath/simd.inc.src

```diff
@@ -2654,6 +2671,188 @@ static NPY_GCC_OPT_3 NPY_GCC_TARGET_@ISA@ void
 #endif
 /**end repeat**/
 
+/*
+ * Vectorized implementation of exp double using AVX512
+ * Reference: Tang, P.T.P., “Table-driven implementation of the
+ *  exponential function in IEEE floating-point
+ *  arithmetic,” ACM Transactions on Mathematical
+ *  Software, vol. 15, pp. 144-157, 1989.
+ * 1) if x > mTH_max or x is INF; return INF (overflow)
+ * 2) if x < mTH_min; return 0.0f (underflow)
+ * 3) if abs(x) < mTH_nearzero; return 1.0f + x
+ * 4) Range reduction:
+ *    x = (32m + j)ln2 / 32 + r; r in [-ln2/64, ln2/64]
+ * 5) exp(r) - 1 is approximated by a polynomial function p(r)
+ *    exp(x) = 2^m(2^(j/32) + 2^(j/32)p(r));
+ */
+#if defined HAVE_ATTRIBUTE_TARGET_AVX512F_WITH_INTRINSICS
+static NPY_GCC_OPT_3 NPY_GCC_TARGET_AVX512F void
+AVX512F_exp_DOUBLE(npy_double * op,
+                npy_double * ip,
+                const npy_intp array_size,
+                const npy_intp steps)
+{
+    /* 
+     * Lookup table for 2^(j/32)
+     * 2^(j/32) ~= EXP_T[2*j] + EXP_T[2*j+1];
+     */
+    static npy_uint64 EXP_T[64] = {
+        0x3FF0000000000000, 0x0000000000000000,
```

Comment:
Perhaps this should be moved to some common header?

Reply:
I've moved the tables into numpy/core/include/numpy/npy_math.h. But this causes a lot of unused warning. Should I create a new header file or move the tables into other header file?

Reply:
Sorry, I missed the comment of mattip, I'll add `if defined HAVE_ATTRIBUTE_TARGET_AVX512F_WITH_INTRINSICS && defined  NPY_HAVE_SSE2_INTRINSICS` in the header file.

Reply:
> Should I create a new header file or move the tables into other header file?

You should not be adding anything to `numpy/core/include`, those are all public headers. I see why you are mislead: there is content in `npy_math.h` which does not belong there. You should create a new header file under `numpy/core/src/umath`. What do you think of the name `numpy/core/src/umath/npy_tables.h`? It would be nice to move the other non-public constants from `npy_math.h` into the new file, I think all the "constansts used in vector ..." can be safely moved.